##1.데이터프레임 만들기

#pd.DataFrame({<컬럼 이름>:<데이터>})
#pd.DataFrame(data=<데이터>, columns=<컬럼이름>)

In [1]:
import pandas as pd

In [ ]:
import numpy as np

dataset = np.array([['kor',70],['math',80]])
dataset

array([['kor', '70'],
       ['math', '80']], dtype='<U21')

In [ ]:
df = pd.DataFrame(dataset,columns=['class','score'])

df

class score
0   kor    70
1  math    80

In [ ]:
df = pd.DataFrame(data=[['kor',70],['math',80]],columns = ['class','score'])

df

class  score
0   kor     70
1  math     80

#이게 가장 좋은듯?

In [ ]:
df = pd.DataFrame({'class':['kor','math'],'score':[70,80]})

df

class  score
0   kor     70
1  math     80

In [ ]:
s = pd.Series({1:'math',2:'kor'},name='class')

s

1    math
2     kor
Name: class, dtype: object

In [ ]:
pd.Series(['math','kor'],name='class')

0    math
1     kor
Name: class, dtype: object

#2.데이터프레임 읽고 저장

pd.read_csv(filepath , na_values='NA', encoding = 'utf8')

CSV파일에서 NULL값이 빈칸이 아닌 어떠한 string값으로 저장된 경우 이를 인식하기 위해 na_values를 사용


na_values로 지정한 값들은 판다스가 자동으로 NA로 인식함

영어 외의 다른 형식의 언어로 작성된 값이 포함되어 있는 경우 해당 언어에 맞는 적절한 형식으로 인코딩

한국어가 포함된 경우 encoding='utf8', encoding='cp949'

In [ ]:
import pandas as pd

filepath = '/content/drive/MyDrive/adpdata/iris.csv'

data = pd.read_csv(filepath,na_values='NA', encoding='utf8')

data

sepal length  sepal width  petal length  petal width          target
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

[150 rows x 5 columns]

In [ ]:
df = pd.DataFrame({'class':['kor','math'],'score':[70,80]})

df

class  score
0   kor     70
1  math     80

(dataset 변수명).to_csv(저장할 파일명, header=True, index=True, encoding='utf8')

to_csv 함수를 사용하여 데이터프레임을 csv파일로 저장

header=True, index=True를 하면 컬럼명이랑 인덱스명을 그대로 인식해서 저장함

한국어가 포함된 데이터이면 반드시 encoding='utf8'을 해야 안깨지고 저장할 수 있다

In [ ]:
df.to_csv('result.csv', header=True, index=True, encoding='utf8')

#3. 데이터프레임 출력

head()와 tail()함수를 이용해 데이터프레임의 처음 또는 마지막 몇줄을 출력하여 간단히 데이터 내용을 확인

기본값은 컬럼명 포함 6줄

In [ ]:
from sklearn.datasets import load_iris

iris = load_iris()
iris = pd.DataFrame(iris.data, columns = iris.feature_names)
iris

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]

pd.set_option('display.max_columns',None)

pd.set_option('display.max_rows',None)

를 수행하면 데이터프레임의 모든 행, 열을 출력할 수 있다.

시간이 많이 소요될 수 있으니 주의해야함

혹은 데이터프레임 전치시켜서 확인도 가능

In [ ]:
pd.set_option('display.max_columns',None)
pd.set_option('display.max_rows',None)

iris

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
5                  5.4               3.9                1.7               0.4
6                  4.6               3.4                1.4               0.3
7                  5.0               3.4                1.5               0.2
8                  4.4               2.9                1.4               0.2
9                  4.9               3.1                1.5               0.1
10                 5.4               3.7                1.5               0.2
11                 4.8               3.4                1.6               0.2
12                 4.8               3.0                1.4               0.1
13                 4.3               3.0                1.1               0.1
14                 5.8               4.0                1.2               0.2
15                 5.7               4.4                1.5               0.4
16                 5.4               3.9                1.3               0.4
17                 5.1               3.5                1.4               0.3
18                 5.7               3.8                1.7               0.3
19                 5.1               3.8                1.5               0.3
20                 5.4               3.4                1.7               0.2
21                 5.1               3.7                1.5               0.4
22                 4.6               3.6                1.0               0.2
23                 5.1               3.3                1.7               0.5
24                 4.8               3.4                1.9               0.2
25                 5.0               3.0                1.6               0.2
26                 5.0               3.4                1.6               0.4
27                 5.2               3.5                1.5               0.2
28                 5.2               3.4                1.4               0.2
29                 4.7               3.2                1.6               0.2
30                 4.8               3.1                1.6               0.2
31                 5.4               3.4                1.5               0.4
32                 5.2               4.1                1.5               0.1
33                 5.5               4.2                1.4               0.2
34                 4.9               3.1                1.5               0.2
35                 5.0               3.2                1.2               0.2
36                 5.5               3.5                1.3               0.2
37                 4.9               3.6                1.4               0.1
38                 4.4               3.0                1.3               0.2
39                 5.1               3.4                1.5               0.2
40                 5.0               3.5                1.3               0.3
41                 4.5               2.3                1.3               0.3
42                 4.4               3.2                1.3               0.2
43                 5.0               3.5                1.6               0.6
44                 5.1               3.8                1.9               0.4
45                 4.8               3.0                1.4               0.3
46                 5.1               3.8                1.6               0.2
47                 4.6               3.2                1.4               0.2
48                 5.3               3.7                1.5               0.2
49                 5.0               3.3                1.4               0.2
50                 7.0     

In [ ]:
iris.T #데이터프레임 전치시키기, #보통 모든 컬럼을 확인하는데 유용함

0    1    2    3    4    5    6    7    8    9    ...  140  \
sepal length (cm)  5.1  4.9  4.7  4.6  5.0  5.4  4.6  5.0  4.4  4.9  ...  6.7   
sepal width (cm)   3.5  3.0  3.2  3.1  3.6  3.9  3.4  3.4  2.9  3.1  ...  3.1   
petal length (cm)  1.4  1.4  1.3  1.5  1.4  1.7  1.4  1.5  1.4  1.5  ...  5.6   
petal width (cm)   0.2  0.2  0.2  0.2  0.2  0.4  0.3  0.2  0.2  0.1  ...  2.4   

                   141  142  143  144  145  146  147  148  149  
sepal length (cm)  6.9  5.8  6.8  6.7  6.7  6.3  6.5  6.2  5.9  
sepal width (cm)   3.1  2.7  3.2  3.3  3.0  2.5  3.0  3.4  3.0  
petal length (cm)  5.1  5.1  5.9  5.7  5.2  5.0  5.2  5.4  5.1  
petal width (cm)   2.3  1.9  2.3  2.5  2.3  1.9  2.0  2.3  1.8  

[4 rows x 150 columns]

In [ ]:
iris.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [ ]:
iris.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

#4. dataframe 요약, 통계정보 확인하기

데이터의 대략적인 내용을 출력해서 확인했다면 데이터프레임의 요약, 통계정보를 확인해야함

info()함수를 통해 총 데이터의 건수와 각 컬럼의 데이터 형식과 결측치의 건수를 확인

수치형 컬럼이 존재하는 경우 describe()함수가 n-percentile분포, 평균, 최대, 최소를 확인할 수 있다

In [ ]:
from sklearn.datasets import load_iris

iris = load_iris()
iris = pd.DataFrame(iris.data, columns = iris.feature_names)

In [ ]:
iris.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


iris는 pandas의 데이터프레임 객체

150개의 행과 4개의 열로 구성

각 열은 sepal length, sepal width, petal length, petal width

Non-Null count에서 볼 수 있듯이 각 칼럼은 결측치를 가지고 있지 않다

모든 칼럼의 데이터타입은 float64

In [ ]:
iris.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)  
count        150.000000  
mean           1.199333  
std            0.762238  
min            0.100000  
25%            0.300000  
50%            1.300000  
75%            1.800000  
max            2.500000

describe()는 수치형 데이터의 개수, 평균, 표준편차, 최솟값, 제1사분위수, 제2사분위수, 제3사분위수, 최댓값을 제공

주로 mean, min, max를 확인

iris의 sepal length의 경우 4.3~7.9의 범위를 가지고 

petal width의 경우 0.1~2.5로 범위 차이가 난다

회귀분석을 수행할 경우 크기가 상대적으로 적은 petal width는 sepal length 같은 숫자가 큰 다른 칼럼들에 비해 분석모델에 미치는 영향력이 적을 수 있다

이는 모델링 전처리 과정에서 변수의 정규화를 수행하는 근거가 된다

#5. index 확인, 추가, 리셋

데이터프레임은 index와 column(속성)을 가지고 있고 이를 통해 인덱스나 컬럼을 수정하거나 출력

인덱스나 컬럼을 수정할 경우 list나 numpy array형식을 사용할 수 있고 기존 데이터와 길이가 같아야함

In [ ]:
import pandas as pd

df = pd.DataFrame({'class':['kor','math'],'score':[70,80]})

df

class  score
0   kor     70
1  math     80

df.index로 데이터프레임 df의 index 열을 출력함

In [ ]:
df.index

RangeIndex(start=0, stop=2, step=1)

In [ ]:
list(df.index)

[0, 1]

df.index = (리스트)로 원하는 인덱스로 변경가능

In [ ]:
df.index = ['A','B']

In [ ]:
df.index

Index(['A', 'B'], dtype='object')

In [ ]:
df

class  score
A   kor     70
B  math     80

set_index()를 사용하면 데이터프레임내의 열을 인덱스로 변경

set_index(keys, drop=True, append=False, inplace=True)

keys: 인덱스로 사용하고자 하는 컬럼의 이름

drop: 인덱스로 세팅한 컬럼을 데이터프레임 내에서 삭제할지 결정

append: 기존에 존재하던 인덱스를 삭제할지 인덱스로 그대로 둘지

inplace: 원본 객체를 변경할지

In [ ]:
df.set_index('class',drop=True, append=False)

score
class       
kor       70
math      80

In [ ]:
df.set_index('class',drop=False,append=False)

class  score
class             
kor     kor     70
math   math     80

In [ ]:
df.set_index('class',drop=True, append=True)

score
  class       
A kor       70
B math      80

In [ ]:
df.set_index('class',drop=True, append=True).index

MultiIndex([('A',  'kor'),
            ('B', 'math')],
           names=[None, 'class'])

inplace를 설정하지않으면 원본 객체를 그대로 둔다

In [ ]:
df.set_index('class',drop=True,append=False)

score
class       
kor       70
math      80

In [ ]:
df

class  score
A   kor     70
B  math     80

In [ ]:
df.set_index('class',drop=True,append=False,inplace=True)

In [ ]:
df

score
class       
kor       70
math      80

reset_index()는 인덱스를 0부터 시작하는 정수로 재설정함

reset_index(drop=False, inplace=False)

drop: 기존 인덱스를 데이터프레임 내에서 삭제할지, 컬럼으로 만들지

inplace: 원본 객체를 변경할지

In [ ]:
df.reset_index(drop=False, inplace=False)

class  score
0   kor     70
1  math     80

In [ ]:
df.reset_index(drop=True, inplace=False)

score
0     70
1     80

#6. 데이터프레임 컬럼명 확인 및 변경

df.index와 비슷하게 df.columns 같이 columns 속성을 사용하여 컬럼명을 확인하거나 변경

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
from sklearn.datasets import load_iris

iris = load_iris()
iris = pd.DataFrame(iris.data, columns = iris.feature_names)
iris.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [ ]:
iris.columns #컬럼명 확인

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

In [ ]:
iris.columns = ['sepal length','sepal width','petal length','petal width']

iris.head()

sepal length  sepal width  petal length  petal width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2

In [ ]:
iris.columns[0] = 'sepal_length' #일부만 바꾸는건 안되는구나

TypeError: ignored

str.replace()함수를 사용하여 컬럼명 내의 특정문자를 다른 문자로 대체

df.columns.str.replace('기존문자','대체할문자')

In [ ]:
iris.columns = iris.columns.str.replace(' ','_')

iris.head()

sepal_length  sepal_width  petal_length  petal_width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2

#7. 데이터프레임 컬럼의 데이터 타입 확인 및 변경

데이터프레임 컬럼은 int, float, bool, datetime, category, object 등 다양한 타입을 가짐

df.dtypes를 통해 컬럼의 타입을 확인할 수 있다.

info()에도 타입 정보를 가지고 있다

In [3]:
import pandas as pd
import numpy as np

In [26]:
from sklearn.datasets import load_iris

iris = load_iris()
iris = pd.DataFrame(iris.data, columns = iris.feature_names)
iris.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [27]:
iris.dtypes

sepal length (cm)    float64
sepal width (cm)     float64
petal length (cm)    float64
petal width (cm)     float64
dtype: object

In [28]:
iris.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


필요에 따라 데이터 타입을 변경해야할 필요가 있다

astype()을 통해 변경 가능

In [30]:
iris.columns = ['sepal_length','sepal_width','petal_length','petal_width']

In [31]:
iris.dtypes

sepal_length    float64
sepal_width     float64
petal_length    float64
petal_width     float64
dtype: object

In [32]:
iris.head()

sepal_length  sepal_width  petal_length  petal_width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2

파이썬에서 다음 줄로 코드를 넘기고 싶을 때 \를 추가하면 다음 줄로 코드를 넘길 수 있다

In [33]:
iris[['sepal_width','petal_length']] = 
iris[['sepal_width','petal_length']].astype('int')

SyntaxError: ignored

In [34]:
iris['sepal_length'] = iris['sepal_width'].astype('int')

iris[['sepal_width','petal_length']] = \ #\추가하여 다음 줄로 넘김
iris[['sepal_width','petal_length']].astype('int')

In [35]:
iris.head()

sepal_length  sepal_width  petal_length  petal_width
0             3            3             1          0.2
1             3            3             1          0.2
2             3            3             1          0.2
3             3            3             1          0.2
4             3            3             1          0.2